### PR TITLE
[kotlin-spring] add reactive behavior via Kotlin coroutines

### PR DESCRIPTION
**Pull request by sylvainmoindron**
### PR checklist

- [X] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [X] Ran the shell script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh`, `./bin/openapi3/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates). Windows batch files can be found in `.\bin\windows\`. If contributing template-only or documentation-only changes which will change sample output, be sure to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) first.
- [x] Filed the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`.
- [x] Copied the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

### Description of the PR
add reactive behavior via kotlin's  coroutines for #2765

@jimschubert , @dr4ke616 @karismann @Zomzog

@botcoder

### CODE REVIEW

File: modules/openapi-generator/src/main/resources/kotlin-spring/libraries/spring-boot/buildGradleKts.mustache [hunk 1]
```diff
@@ -23,18 +23,26 @@ tasks.withType<KotlinCompile> {
...
+	val kotlinxCoroutinesVersion="1.2.0"
     compile("org.jetbrains.kotlin:kotlin-stdlib-jdk8")
     compile("org.jetbrains.kotlin:kotlin-reflect")
+{{^reactive}}
     compile("org.springframework.boot:spring-boot-starter-web")
+{{/reactive}}
+{{#reactive}}
+	compile("org.springframework.boot:spring-boot-starter-webflux")
+	compile("org.jetbrains.kotlinx:kotlinx-coroutines-core:$kotlinxCoroutinesVersion")
+	compile("org.jetbrains.kotlinx:skotlinx-coroutines-reactor$kotlinxCoroutinesVersion")
```
Comment (Zomzog) [inline modules/openapi-generator/src/main/resources/kotlin-spring/libraries/spring-boot/buildGradleKts.mustache]: There is a missing `:`  at this line.

**Comment by sylvainmoindron** [comment]
with openapi in general (not related to this PR) i don't understand why pathParam, queryParm and bodyParam are not handled the same way for data type

**Comment by sylvainmoindron** [comment]
@jimschubert @Zomzog, in the classe UserApi methode createUsersWithArrayInput in my mustache template I use
`Flow<{{{baseType}}}>` and it's converted to `Flow<List>` instead of ` Flow<User>` what did i do wrong ?

**Comment by wing328** [comment]
>Flow<{{{baseType}}}> and it's converted to Flow<List> instead of Flow<User> what did i do wrong ?

To be clear, the input parameter is an array/list of User (model). Are you saying `Flow<User>` is the correct syntax to describe such input parameter?

**Comment by sylvainmoindron** [comment]
@wing328 yes `Flow<User>` is the correct type for a list of user in reactive mode

**Comment by Zomzog** [comment]
I tried your PR localy with 
``` 
 /ponies:
    summary: "Everything for ponies"
    get:
      summary: "List"
      operationId: "list"
      responses:
        200:
          description: "Status 200"
          content:
            application/json:
              schema:
                type: "array"
                items:
                  $ref: "#/components/schemas/Pony"
```
and I've got `ResponseEntity<Flow<Pony>>` so I'm not sure to understand your issue.

**Comment by Zomzog** [comment]
There is an issue with type resolution. There is the same one with Spring, I will try to understand why it after working hours to help you.

**Comment by sylvainmoindron** [comment]
@Zomzog I think i found the baseType bug

**Comment by sylvainmoindron** [comment]
do i need to run ` bin/utils/ensure-up-to-date` on local before comit since i modified a common element ?

**Comment by wing328** [comment]
@sylvainmoindron thanks for fixing the bug in the default codegen.

If no further feedback on this PR, I'll merge it tomorrow (Thur)

**Comment by jimschubert** [comment]
I've done a quick look over this PR and I think it looks good.

**Comment by jimschubert** [comment]
Sorry for the delay in merge. We hit a snag with the gradle plugin deployment in the 4.0.1 patch release, which pushed this merge back a few days.

**Comment by sylvainmoindron** [comment]
does your response is a single item or a list ?

File: modules/openapi-generator/src/main/resources/kotlin-spring/api.mustache [hunk 2]
```diff
@@ -68,12 +72,12 @@ class {{classname}}Controller({{#serviceInterface}}@Autowired(required = true) v
...
-            produces = "{{{vendorExtensions.x-accepts}}}", {{/hasProduces}}{{#hasConsumes}}
+            produces = "{{{vendorExtensions.x-accepts}}}",{{/hasProduces}}{{#hasConsumes}}
             consumes = "{{{vendorExtensions.x-contentType}}}",{{/hasConsumes}}{{/singleContentTypes}}{{^singleContentTypes}}{{#hasProduces}}
             produces = [{{#produces}}"{{{mediaType}}}"{{#hasMore}}, {{/hasMore}}{{/produces}}], {{/hasProduces}}{{#hasConsumes}}
             consumes = [{{#consumes}}"{{{mediaType}}}"{{#hasMore}}, {{/hasMore}}{{/consumes}}],{{/hasConsumes}}{{/singleContentTypes}}
             method = [RequestMethod.{{httpMethod}}])
-    fun {{operationId}}({{#allParams}}{{>queryParams}}{{>pathParams}}{{>headerParams}}{{>bodyParams}}{{>formParams}}{{#hasMore}},{{/hasMore}}{{/allParams}}): ResponseEntity<{{>returnTypes}}> {
+    {{#reactive}}{{^isListContainer}}suspend {{/isListContainer}}{{/reactive}}fun {{operationId}}({{#allParams}}{{>queryParams}}{{>pathParams}}{{>headerParams}}{{>bodyParams}}{{>formParams}}{{#hasMore}},{{/hasMore}}{{/allParams}}): ResponseEntity<{{>returnTypes}}> {
```
[the remark this answers is not in the document]
Comment (sylvainmoindron) [reply modules/openapi-generator/src/main/resources/kotlin-spring/api.mustache:80]: does your response is a single item or a list ?

[the remark this answers is not in the document]
Comment (elijah-pl) [reply modules/openapi-generator/src/main/resources/kotlin-spring/api.mustache:80]: @sylvainmoindron A list
```yaml
paths:
  /something:
    get:
      operationId: getSomething
      responses:
        "200":
          description: OK
          content:
            application/json:
              schema:
                type: array
                items:
                  $ref: '#/components/schemas/Something'
```
Generated method looks like
```kotlin
fun getSomething(): ResponseEntity<Flow<Something>>
```
And I want it to be
```kotlin
suspend fun getSomething(): ResponseEntity<List<Something>>
```

[the remark this answers is not in the document]
Comment (sylvainmoindron) [reply modules/openapi-generator/src/main/resources/kotlin-spring/api.mustache:80]: if your result is a list it's nomal that the method return a flow. flow are member of coroutine.
in your flow you can call suspend method or methos the return a flow.
when you define a flow with `flow{ action here }`, your action are in a coroutineScope.

[the remark this answers is not in the document]
Comment (sylvainmoindron) [reply modules/openapi-generator/src/main/resources/kotlin-spring/api.mustache:80]: if you want single List you can define your return as an object SometingList with contain your list

**Pull request merged**